### PR TITLE
JDK16 Changes

### PR DESCRIPTION
**Pull request by babsingh**
1. Add `String.getBytes` for JDK16.

2. Add `Lookup.ORIGINAL` and update `defineHiddenClassWithClassData` for JDK16.

3. Add implementation of `JavaLangAccess.addExports` for JDK16.

Closes: #11312

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

### CODE REVIEW

**Comment by pshipton** [comment]
jenkins compile zlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

**Comment by pshipton** [comment]
jenkins test sanity zlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

**Comment by babsingh** [comment]
@pshipton Made a minor correction. The PR builds would need to be relaunched.

**Comment by tajila** [comment]
jenkins test sanity zlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

**Comment by tajila** [comment]
what change was made?

**Comment by babsingh** [comment]
> what change was made?

fixed grammar in `String.getBytes`'s description.

**Comment by pshipton** [comment]
FYI new builds aren't necessarily required for changes to comments. The original builds continue on even if you push more changes and they are no longer shown in the PR.

test_insert12 is failing in JCL_Test, which means we won't automatically promote jdknext even if these changes are merged.

**Comment by pshipton** [comment]
@tajila fyi if you look at the `forced-push` link in `babsingh force-pushed the babsingh:jdk16_comp branch from 277aa79 to a7b6eaf 2 hours ago` above, you can see the changes
https://github.com/eclipse/openj9/compare/277aa796c83429afca607c280c1d68aa89051394..a7b6eaf0b7d83ff02fb1644519940e42d15c438e

**Comment by keithc-ca** [comment]
A recent addition to `JavaLangAccess` (`void addExports(Module module, String packageName)`) must be implemented.

**Comment by babsingh** [comment]
Update:
1. I will temporarily disable `test_insert12` and `test_insert3` to enable jdknext builds. Investigating and fixing these failures will take some time. 
2. I will also add `JavaLangAccess.addExports` to this PR.

**Comment by keithc-ca** [comment]
Perhaps the failure in `test_insert12` was fixed by openjdk?
* 8257511: JDK-8254082 brings regression to AbstractStringBuilder.insert(int dstOffset, CharSequence s, int start, int end)

See https://github.com/ibmruntimes/openj9-openjdk-jdk/blame/11ff3afa7597ba8742399ddf387ab5e17a9929f8/src/java.base/share/classes/java/lang/AbstractStringBuilder.java#L1720

**Comment by babsingh** [comment]
> Perhaps the failure in test_insert12 was fixed by openjdk?

Rebuilding with the latest source to confirm.

**Comment by babsingh** [comment]
The PR has been updated to include `JavaLangAccess.addExports`.

The OpenJDK commit mentioned in https://github.com/eclipse/openj9/pull/11322#issuecomment-737361732 fixes `test_insert12` and `test_insert3`.

**Comment by pshipton** [comment]
jenkins test sanity zlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#openj9-staging

File: jcl/src/java.base/share/classes/java/lang/String.java [hunk 1]
```diff
@@ -113,6 +113,33 @@ byte coder() {
...
+	 * Copy bytes from value starting at srcIndex into the bytes array starting at
+	 * destIndex. No range checking is needed. Caller ensures bytes is in UTF16.
+	 * 
+	 * @param bytes copy destination
+	 * @param srcIndex index into value
+	 * @param destIndex index into bytes
+	 * @param coder LATIN1 or UTF16
+	 * @param length the number of elements to copy
+	 * 
+	 * @return void
```
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/String.java:127]: I don't think this is appropriate: please remove.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 2]
```diff
@@ -162,14 +162,20 @@
 		 */
 		public static final int UNCONDITIONAL = 0x20;
 		/*[ENDIF] Sidecar19-SE*/
-		
-		static final int INTERNAL_PRIVILEGED = 0x40;
 
-		/*[IF Sidecar19-SE-OpenJ9]
+		static final int INTERNAL_PRIVILEGED = 0x80;
```
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:166]: What is affected by changing the value of `INTERNAL_PRIVILEGED`?

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/job/PullRequest-OpenJ9/4984/

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 2 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:166]: It is used to set `Lookup.accessMode`. Based upon its current usage, nothing should be affected based by changing its value.

File: jcl/src/java.base/share/classes/java/lang/String.java (hunk 1 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/String.java]: It has been removed.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 2 above)
Comment (keithc-ca) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:166]: Please confirm that comment applies to all jdk versions (8, 11, 15, next).

**Comment by pshipton** [comment]
PR testing passed. The only change was to a comment so it's still valid.

**Review by pshipton — APPROVED** [review]

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 2 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:166]: Yes, it does apply to all jdk versions (8, 11, 15, next).

Comment (keithc-ca) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:166]: Thanks.

**Review by keithc-ca — APPROVED** [review]

**Comment by pshipton** [comment]
@babsingh Seems we should have tested jdk15, as it no longer compiles with this change. The old definition of defineHiddenClassWithClassData() didn't get removed and now there are two.
```
22:44:42  /home/jenkins/workspace/Build_JDK15_ppc64_aix_Nightly/build/aix-ppc64-server-release/support/j9jcl/java.base/share/classes/java/lang/invoke/MethodHandles.java:2075: error: method defineHiddenClassWithClassData(byte[],Object,ClassOption...) is already defined in class Lookup
22:44:42  		Lookup defineHiddenClassWithClassData(byte[] bytes, Object classData, ClassOption... classOptions) throws IllegalAccessException {
```

**Pull request merged**